### PR TITLE
chore: add new interfaces for Assets

### PR DESCRIPTION
**Pull request by skinny85**
In V2, we want to get rid of the `@aws-cdk/assets` module, as it's considered deprecated in V1.
Unfortunately, the module contains an interface, `CopyOptions`, that is used, through interface inheritance, in the public API of many stable CDK modules like ECS, Lambda, ECR, CodeBuild, etc.

While we have a `CopyOptions` interface in `@aws-cdk/core`, it unfortunately shares the same name, `follow`, with the property in the "old" `CopyOptions`. But the two different `follow` properties have different types.

For that reason, if we're going to remove the "old" `CopyOptions` using JSII's "strip deprecated" option, we can't use the "new" `CopyOptions` in the inheritance hierarchy alongside the "old" `CopyOptions`, as the two definitions of `follow` would conflict.

Because of that, create a new `FileCopyOptions` interface which renames the `follow` property to `followSymlinks`. Also add a `FileFingerprintOptions` interface that does a similar trick to the `FingerprintOptions` interface (which extends `CopyOptions`), which is used in the public API of modules that use Docker assets.

Also extract a few module-private interfaces to avoid duplication of properties between all of these interfaces.

After this change, an interface from one of the non-deprecated assets libraries (S3 or ECR) using `FileOptions` from `@aws-cdk/assets` will look like this:

```ts
// this is in @aws-cdk/aws-s3-assets
export interface AssetOptions extends assets.CopyOptions, cdk.FileCopyOptions, cdk.AssetOptions {
  // ...
```

Then, when we enable stripping the deprecated elements using JSII on the V2 branch, this will be turned to:

```ts
export interface AssetOptions extends cdk.FileCopyOptions, cdk.AssetOptions {
  // ...
```

Allowing us to deprecate the `@aws-cdk/assets` module, and not ship it with `aws-cdk-lib`.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-ecr-assets/lib/image-asset.ts
Comment (nija-at) [inline packages/@aws-cdk/aws-ecr-assets/lib/image-asset.ts:12]: I assume changes to this file are not related to your PR description, yes?

File: packages/@aws-cdk/core/lib/fs/options.ts
Comment (nija-at) [inline packages/@aws-cdk/core/lib/fs/options.ts:92]: Do we need both of these or can we deprecate the former?

**Review by nija-at — COMMENTED** [review]
cc @eladb who is more familiar with this part of the CDK.

File: packages/@aws-cdk/aws-ecr-assets/lib/image-asset.ts
Comment (skinny85) [reply packages/@aws-cdk/aws-ecr-assets/lib/image-asset.ts:12]: I don't understand what you're asking, sorry. They are very much related (in fact, the description explicitly calls out `FingerprintOptions`).

File: packages/@aws-cdk/core/lib/fs/options.ts
Comment (skinny85) [reply packages/@aws-cdk/core/lib/fs/options.ts:92]: Again, sorry, I don't understand what you're asking here (deprecate which interface?).

Comment (nija-at) [reply packages/@aws-cdk/core/lib/fs/options.ts:92]: You don't have to be sorry. It's ok.

I was asking if we should  `CopyOptions` and `FileCopyOptions`. They look exactly the same.

Comment (skinny85) [reply packages/@aws-cdk/core/lib/fs/options.ts:92]: Sure, we can deprecate `CopyOptions`. It's used in a bunch of public classes, so it will require changes to augment `CopyOptions` with `FileCopyOptions` in all of these usages, and it's not linked to deprecating `@aws-cdk/assets` directly.

Let me know if you want to make this part of this PR, or tackle it separately after this is merged.

Comment (nija-at) [reply packages/@aws-cdk/core/lib/fs/options.ts:92]: Can you get @eladb's eyes on this? He owns and has built this area and it'd best for him to take a look at this PR and assess this change.

I'll defer to him on what the right thing here is.

File: packages/@aws-cdk/assets/lib/fs/options.ts
Comment (eladb) [inline packages/@aws-cdk/assets/lib/fs/options.ts:13]: Where is `followSymlinks` defined?

File: packages/@aws-cdk/core/lib/fs/options.ts
Comment (eladb) [inline packages/@aws-cdk/core/lib/fs/options.ts:78]:
```suggestion
 * Options applied when copying directories
```

(this is a general utility used by asset staging but can technically be used for other purposes)

**Review by eladb — COMMENTED** [review]
LGTM, but before I approve, can you please add to the PR description the final state that we will have in V2 after the all the deprecations are removed from the library?

File: packages/@aws-cdk/assets/lib/fs/options.ts
Comment (skinny85) [reply packages/@aws-cdk/assets/lib/fs/options.ts:13]: It's defined in `FileCopyOptions`.

**Comment by skinny85** [comment]
> LGTM, but before I approve, can you please add to the PR description the final state that we will have in V2 after the all the deprecations are removed from the library?

@eladb done! Let me know if the description makes sense now.

**Review by eladb — APPROVED** [review]

**Pull request merged**
